### ISSUE TITLE
Add API key dashboard

### ISSUE THREAD

**Issue by ericmj** [hexpm/hexpm#668]
Lists all API keys for the user and gives the ability to revoke or generate new ones.

**Comment by boddhisattva** [comment]
@ericmj I'd like to give this a shot.

**Comment by ericmj** [comment]
Go ahead. Thank you.

**Comment by boddhisattva** [comment]
Thanks Eric!

**Comment by boddhisattva** [comment]
@ericmj and other collaborators on this project,

I'm new to the codebase and I have a few questions with regard to this requirement. Can you please see if you could answer the below?

1. Can you please elaborate more on what is the general purpose that these API keys serve?
2. I was thinking we could probably have a `dashboard/keys` page to have User specific keys. Is there any other specific section of the website where you think you'd want to have this API key dashboard?
3. I was exploring the codebase and based on what I researched so far it looks to me that we need to implement a non API version of the `key_controller.ex` with the appropriate functions. Can you please let me know if I'm thinking in the right direction?
4. Also can we consider to breakdown the requirements into 2 parts, one each for the below:
1. Lists all API keys for the user
2. Generate new keys and revoke existing ones

Thank you.

**Comment by boddhisattva** [comment]
Thank you for answering my questions @wojtekmach :)

**Comment by boddhisattva** [comment]
Dear Team, 

Can one please confirm what are all the contents we'd like to show as part of `dashboard/keys` page with regard to the first part of the requirement which is got to do with just listing keys ? Currently, as part of the dashboard keys page, I'm displaying the User key related `name` and `secret_first` attributes. Is there anything you suggest we could add/remove in terms of the User key related attributes displayed on this page? PFB a screenshot of how things look so far:

<img width="1250" alt="screen shot 2018-05-03 at 5 47 02 pm" src="https://user-images.githubusercontent.com/655963/39575982-10fea7fc-4efa-11e8-8878-3a494a421160.png">

Also, it would be great if one could throw some light on what is the difference between the following Key related attributes - `secret_first`, `secret_second` and `user_secret` . 

Lastly, please let me know if you have any other feedback based on the screenshot shared above.

Thank you :)

**Comment by ericmj** [comment]
It looks good but I don't think we should show the secret values because they are of no use to the user. The `user_secret` is only never stored and is the value returned to the user. The `secret_first` and `secret_second` are the stored values and they are calculated from the `user_secret` each time we test a key.

We also want to add a revoke button and show the permissions for each key in the table. In the future we are also going to add this information to the table https://github.com/hexpm/hexpm/issues/668.

**Comment by boddhisattva** [comment]
Thanks for your prompt response and for answering my questions @ericmj!. Since you've mentioned that the secret values would be of no use to the user, I'll remove them.

Are we looking to display the permissions related to each key for the PR related to listing keys or would that be another requirement altogether? I'm not quite sure about what details are needed to be displayed as part of permissions for each key, hence the question. I'm planning to add the revoke key and generate new key functionality as part of another PR based on the discussion above. Hope that's ok.

**Comment by ericmj** [comment]
It is perfectly fine to add the permissions in a separate PR, it is usually better to get something minimal ready and iterate from there.

**Comment by boddhisattva** [comment]
Thanks @ericmj . I have just one more thing to clarify as part of the PR opened [here](https://github.com/hexpm/hexpm/pull/669). Can you please clarify the role of `authing_key` as part of the code [here](https://github.com/hexpm/hexpm/blob/master/lib/hexpm/web/controllers/api/key_controller.ex#L14)?

**Comment by ericmj** [comment]
It's the key used to authorize the current request. Since we don't store the same key that the user does this is the only way of telling, for example, the name of a given key.

**Comment by boddhisattva** [comment]
Thanks for clarifying this Eric.

**Comment by boddhisattva** [comment]
Dear Team,

Just an update from my end: 

I have implemented the revoking keys related functionality and currently have it in my localbox. I was thinking of sending a WIP PR once [this PR](https://github.com/hexpm/hexpm/pull/669) is merged. Alternatively, I can update my [current PR](https://github.com/hexpm/hexpm/pull/669) if you'd like me to do so such that it will also have the revoke key functionality and then we can have the other PR just to implement the Generate new keys for the user. Please let me know what works best for you given the updates mentioned above.

Below is a screenshot of how things look so far(on my local box and not based on the [current PR](https://github.com/hexpm/hexpm/pull/669)) that has combined changes of - List User API Keys(with the latest feedback on the PR incorporated with regard to using bootstrap CSS) and add the revoke keys functionality:

<img width="1269" alt="screen shot 2018-05-04 at 7 13 54 pm" src="https://user-images.githubusercontent.com/655963/39631299-6fa25d82-4fcf-11e8-947a-67c3c67ce664.png">

I'd like to just confirm one thing at this point, for the functionality related to 'Generate keys', we'd like to  simply have a text field that would take the key name and we'd like to generate a new key for the user based on only that info right ? I'm planning to implement something on the `/dashboard/keys` page itself to generate a new key and I'm also planning to implement something like the below screenshot(taken from `/dashboard/email` page) in order to generate new user keys. Kindly let me know if you'd like the generate keys to be implemented in another way.

<img width="725" alt="screen shot 2018-05-04 at 7 17 51 pm" src="https://user-images.githubusercontent.com/655963/39631565-1bc9c050-4fd0-11e8-8f12-0e7e6cd93673.png">


Thank you :)

**Comment by boddhisattva** [comment]
Sure @ericmj, I've updated the PR. You can see a screenshot of how the `/dashboard/keys` page now looks based on the latest changes suggested above to use `table-striped` related Bootstrap CSS [here](https://github.com/hexpm/hexpm/pull/669#issue-185725160).

Thank you for confirming on the approach I could use with regard to implementing the Generate new keys functionality.

**Comment by boddhisattva** [comment]
Dear Team, I've created a new PR [here](https://github.com/hexpm/hexpm/pull/670#issue-186146287) to add the functionality related to revoke and generate user API keys on web. Please review when feasible. Thank you.

**Comment by boddhisattva** [comment]
Hi @ericmj , I've made the changes to the [List User API Keys on web](https://github.com/hexpm/hexpm/pull/669#pullrequestreview-117345494) PR based on your review comment [here](https://github.com/hexpm/hexpm/pull/669#pullrequestreview-117345494). I've also submitted a PR on the Add and Revoke Keys related functionality [here](https://github.com/hexpm/hexpm/pull/670). Can you please take a look when you get some time ? 

Dear Team, 

I've asked a question [here](https://github.com/hexpm/hexpm/pull/670#issuecomment-387005489) about the order in which you'd like to have the changes merged with regard to the requirements implemented as part of this issue, it would be great if you could answer this when feasible.

Thank you.
